File: src/main/java/com/planbase/pdf/layoutmanager/Cell.java
// Copyright 2013-03-03 PlanBase Inc. & Glen Peterson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.planbase.pdf.layoutmanager;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 A styled table cell or layout block with a pre-set horizontal width.  Vertical height is calculated 
 based on how the content is rendered with regard to line-breaks and page-breaks.
 */
public class Cell implements Renderable {

    // These are limits of the cell, not the contents.
    private final CellStyle cellStyle;
    private final double width;

    // A list of the contents.  It's pretty limiting to have one item per row.
    private final List<Renderable> rows;

    private final Map<Double,PreCalcRows> preCalcRows = new HashMap<Double,PreCalcRows>(0);

    private static class PreCalcRow {
        Renderable row;
        Dim blockDim;
        public static PreCalcRow of(Renderable r, Dim d) {
            PreCalcRow pcr = new PreCalcRow(); pcr.row = r; pcr.blockDim = d; return pcr;
        }
    }

    private static class PreCalcRows {
        List<PreCalcRow> rows = new ArrayList<PreCalcRow>(1);
        Dim blockDim;
    }

    private Cell(CellStyle cs, double w, List<Renderable> rs) {
        if (w < 0) {
            throw new IllegalArgumentException("A cell cannot have a negative width");
        }
//        for (Renderable r : rs) {
//            if (r == null) {
//                throw new IllegalArgumentException("How am I supposed to render a null?");
//            }
//        }
        cellStyle = cs; width = w; rows = rs;
    }

    /**
     Creates a new cell.

     @param w the width (height will be calculated based on how objects can be rendered within this
         width).
     @param cs the cell style
     @return a cell suitable for rendering.
     */
    public static Cell of(CellStyle cs, double w) { //, final Object... r) {
        return new Cell(cs, w, Collections.<Renderable>emptyList());
//                        (r == null) ? Collections.emptyList()
//                                    : Arrays.asList(r));
    }

    // Simple case of a single styled String
    public static Cell of(CellStyle cs, double w, TextStyle ts, String s) {
        List<Renderable> ls = new ArrayList<Renderable>(1);
        ls.add(Text.of(ts, s));
        return new Cell(cs, w, ls);
    }

    // Simple case of a single styled String
    public static Cell of(CellStyle cs, double w, Text t) {
        List<Renderable> ls = new ArrayList<Renderable>(1);
        ls.add(t);
        return new Cell(cs, w, ls);
    }

    public static Cell of(CellStyle cs, double w, ScaledJpeg j) {
        List<Renderable> ls = new ArrayList<Renderable>(1);
        ls.add(j);
        return new Cell(cs, w, ls);
    }

    public static Cell of(CellStyle cs, double w, Renderable r) {
        List<Renderable> ls = new ArrayList<Renderable>(1);
        ls.add(r);
        return new Cell(cs, w, ls);
    }

    public static Cell of(CellStyle cs, double w, List<Renderable> ls) {
        return new Cell(cs, w, ls);
    }

    // Simple case of a single styled String
    public static Cell of(CellStyle cs, double w, Cell c) {
        List<Renderable> ls = new ArrayList<Renderable>(1);
        ls.add(c);
        return new Cell(cs, w, ls);
    }

    public CellStyle cellStyle() { return cellStyle; }
    // public BorderStyle border() { return borderStyle; }
    public double width() { return width; }

    private void calcDimensionsForReal(final double maxWidth) {
        PreCalcRows pcrs = new PreCalcRows();
        Dim blockDim = Dim.ZERO;
        Padding padding = cellStyle.padding();
        double innerWidth = maxWidth;
        if (padding != null) {
            innerWidth -= (padding.left() + padding.right());
        }
        for (Renderable row : rows) {
            Dim rowDim = (row == null) ? Dim.ZERO : row.calcDimensions(innerWidth);
            blockDim = Dim.of(Math.max(blockDim.getWidth(), rowDim.getWidth()),
                              blockDim.getHeight() + rowDim.getHeight());
//            System.out.println("\trow = " + row);
//            System.out.println("\trowDim = " + rowDim);
//            System.out.println("\tactualDim = " + actualDim);
            pcrs.rows.add(PreCalcRow.of(row, rowDim));
        }
        pcrs.blockDim = blockDim;
        preCalcRows.put(maxWidth, pcrs);
    }

    private PreCalcRows ensurePreCalcRows(final double maxWidth) {
        PreCalcRows pcr = preCalcRows.get(maxWidth);
        if (pcr == null) {
            calcDimensionsForReal(maxWidth);
            pcr = preCalcRows.get(maxWidth);
        }
        return pcr;
    }

    /** {@inheritDoc} */
    @Override public Dim calcDimensions(final double maxWidth) {
        // I think zero or negative width cells might be OK to ignore.  I'd like to try to make
        // Text.calcDimensionsForReal() handle this situation before throwing an error here.
//        if (maxWidth < 0) {
//            throw new IllegalArgumentException("maxWidth must be positive, not " + maxWidth);
//        }
        Dim blockDim = ensurePreCalcRows(maxWidth).blockDim;
        return ((cellStyle.padding() == null) ? blockDim : cellStyle.padding().addTo(blockDim));
//        System.out.println("Cell.calcDimensions(" + maxWidth + ") blockDim=" + blockDim +
//                           " returns " + ret);
    }

    /*
    Renders item and all child-items with given width and returns the x-y pair of the
    lower-right-hand corner of the last line (e.g. of text).

    {@inheritDoc}
    */
    @Override public Coord render(LogicalPage lp, Coord outerTopLeft,
                                  final Dim outerDimensions, boolean allPages) {
//        System.out.println("Cell.render(" + this.toString());
//        new Exception().printStackTrace();

        double maxWidth = outerDimensions.getWidth();
        PreCalcRows pcrs = ensurePreCalcRows(maxWidth);
        final Padding padding = cellStyle.padding();
        // XyDim outerDimensions = padding.addTo(pcrs.blockDim);

        // Draw background first (if necessary) so that everything else ends up on top of it.
        if (cellStyle.bgColor() != null) {
//            System.out.println("\tCell.render calling putRect...");
            lp.putRect(outerTopLeft, outerDimensions, cellStyle.bgColor());
//            System.out.println("\tCell.render back from putRect");
        }

        // Draw contents over background, but under border
        Coord innerTopLeft;
        final Dim innerDimensions;
        if (padding == null) {
            innerTopLeft = outerTopLeft;
            innerDimensions = outerDimensions;
        } else {
//            System.out.println("\tCell.render outerTopLeft before padding=" + outerTopLeft);
            innerTopLeft = padding.applyTopLeft(outerTopLeft);
//            System.out.println("\tCell.render innerTopLeft after padding=" + innerTopLeft);
            innerDimensions = padding.subtractFrom(outerDimensions);
        }
        Dim wrappedBlockDim = pcrs.blockDim;
//        System.out.println("\tCell.render cellStyle.align()=" + cellStyle.align());
//        System.out.println("\tCell.render outerDimensions=" + outerDimensions);
//        System.out.println("\tCell.render padding=" + padding);
//        System.out.println("\tCell.render innerDimensions=" + innerDimensions);
//        System.out.println("\tCell.render wrappedBlockDim=" + wrappedBlockDim);
        Padding alignPad = cellStyle.align().calcPadding(innerDimensions, wrappedBlockDim);
//        System.out.println("\tCell.render alignPad=" + alignPad);
        if (alignPad != null) {
            innerTopLeft = Coord.of(innerTopLeft.getX() + alignPad.left(),
                                    innerTopLeft.getY() - alignPad.top());
        }

        Coord outerLowerRight = innerTopLeft;
        for (int i = 0; i < rows.size(); i++) {
            Renderable row = rows.get(i);
            if (row == null) {
                continue;
            }
            PreCalcRow pcr = pcrs.rows.get(i);
            double rowXOffset = cellStyle.align().leftOffset(wrappedBlockDim.getWidth(), pcr.blockDim.getWidth());
            outerLowerRight = row.render(lp,
                                         innerTopLeft.withX(innerTopLeft.getX() + rowXOffset),
                                         pcr.blockDim, allPages);
            innerTopLeft = outerLowerRight.withX(innerTopLeft.getX());
        }

        // Draw border last to cover anything that touches it?
        BorderStyle border = cellStyle.borderStyle();
        if (border != null) {
            double origX = outerTopLeft.getX();
            double origY = outerTopLeft.getY();
            double rightX = outerTopLeft.getX() + outerDimensions.getWidth();
            double bottomY = outerTopLeft.getY() - outerDimensions.getHeight();
            // Like CSS it's listed Top, Right, Bottom, left
            if (border.top() != null) {
                lp.putLine(origX, origY, rightX, origY, border.top());
            }
            if (border.right() != null) {
                lp.putLine(rightX, origY, rightX, bottomY, border.right());
            }
            if (border.bottom() != null) {
                lp.putLine(origX, bottomY, rightX, bottomY, border.bottom());
            }
            if (border.left() != null) {
                lp.putLine(origX, origY, origX, bottomY, border.left());
            }
        }

        return outerLowerRight;
    }

    public static Builder builder(CellStyle cellStyle, double width) {
        return new Builder(cellStyle, width);
    }

    // Replaced with TableRow.CellBuilder.of()
//    /**
//     Be careful when adding multiple cell builders at once because the cell size is based upon
//     a pointer into the list of cell sizes.  That pointer gets incremented each time a cell is
//     added, not each time nextCellSize() is called.  Is this a bug?  Or would fixing it create
//     too many other bugs?
//     @param trb
//     @return
//     */
//    public static Builder builder(TableRowBuilder trb) {
//        Builder b = new Builder(trb.cellStyle(), trb.nextCellSize()).textStyle(trb.textStyle());
//        b.trb = trb;
//        return b;
//    }

    /**
     * A mutable Builder for somewhat less mutable cells.
     */
    public static class Builder implements CellBuilder {
        private final double width;
        private CellStyle cellStyle;
        private final List<Renderable> rows = new ArrayList<Renderable>();
        private TextStyle textStyle;

        private Builder(CellStyle cs, double w) { width = w; cellStyle = cs; }

        // Is this necessary?
//        public Builder width(double w) { width = w; return this; }

        /** {@inheritDoc} */
        @Override public Builder cellStyle(CellStyle cs) { cellStyle = cs; return this;}

        /** {@inheritDoc} */
        @Override public Builder align(CellStyle.Align align) {
            cellStyle = cellStyle.align(align); return this;
        }

        /** {@inheritDoc} */
        @Override public Builder textStyle(TextStyle x) { textStyle = x; return this; }

        /** {@inheritDoc} */
        @Override public Builder add(Renderable rs) {
            Collections.addAll(rows, rs); return this;
        }

        /** {@inheritDoc} */
        @Override public Builder addAll(Collection<? extends Renderable> js) {
            if (js != null) { rows.addAll(js); } return this;
        }

        /** {@inheritDoc} */
        @Override public Builder add(TextStyle ts, Iterable<String> ls) {
            if (ls != null) {
                for (String s : ls) {
                    rows.add(Text.of(ts, s));
                }
            }
            return this;
        }

        /** {@inheritDoc} */
        @Override public Builder addStrs(String... ss) {
            if (textStyle == null) {
                throw new IllegalStateException("Must set a default text style before adding raw strings");
            }
            for (String s : ss) {
                rows.add(Text.of(textStyle, s));
            }
            return this;
        }
//        public Builder add(Cell c) { rows.add(c); return this; }

        public Cell build() { return new Cell(cellStyle, width, rows); }

        /** {@inheritDoc} */
        @Override  public double width() { return width; }

// Replaced with TableRow.CellBuilder.buildCell()
//        public TableRowBuilder buildCell() {
//            Cell c = new Cell(cellStyle, width, rows);
//            return trb.addCell(c);
//        }

        /** {@inheritDoc} */
        @Override public String toString() {
            StringBuilder sB = new StringBuilder("Cell.Builder(").append(cellStyle).append(" width=")
                    .append(width).append(" rows=[");

            for (int i = 0; (i < rows.size()) && (i < 3); i++) {
                if (i > 0) { sB.append(" "); }
                sB.append(rows.get(i));
            }
            return sB.append("])").toString();
        }
    }

    /** {@inheritDoc} */
    @Override public String toString() {
        StringBuilder sB = new StringBuilder("Cell(").append(cellStyle).append(" width=")
                .append(width).append(" rows=[");

        for (int i = 0; (i < rows.size()) && (i < 3); i++) {
            if (i > 0) { sB.append(" "); }
            sB.append(rows.get(i));
        }
        return sB.append("])").toString();
    }
}

File: src/main/java/com/planbase/pdf/layoutmanager/Cell.kt
// Copyright 2013-03-03 PlanBase Inc.
//
// This file is part of PdfLayoutMgr2
//
// PdfLayoutMgr is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// PdfLayoutMgr is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with PdfLayoutMgr.  If not, see <https://www.gnu.org/licenses/agpl-3.0.en.html>.
//
// If you wish to use this code with proprietary software,
// contact PlanBase Inc. <https://planbase.com> to purchase a commercial license.

package com.planbase.pdf.layoutmanager

import java.util.ArrayList
import java.util.HashMap

import java.lang.Math.max

/**
 A styled table cell or layout block with a pre-set horizontal width.  Vertical height is calculated
 based on how the content is rendered with regard to line-breaks and page-breaks.
 */
data class Cell(val cellStyle: CellStyle = CellStyle.DEFAULT, // contents can override this style
                val width: Float,
                // A list of the contents.  It's pretty limiting to have one item per row.
                private val contents: List<Renderable>) : Renderable {

    private constructor(b:Builder) : this(b.cellStyle ?: CellStyle.DEFAULT, b.width, b.rows)

    // Caches XyDims for all content lines, indexed by desired width (we only have to lay-out again
    // when the width changes.
    private val widthCache = HashMap<Float, PreCalcLines>(0)

    // A cache for all pre-calculated lines.
    class PreCalcLines(val lines: List<Line> = ArrayList(1),
                       val totalDim: XyDim)

    init {
        if (width < 0) {
            throw IllegalArgumentException("A cell cannot have a negative width")
        }
    }

    private fun calcDimensionsForReal(maxWidth: Float) {
        val padding = cellStyle.padding
        var innerWidth = maxWidth
        if (padding != null) {
            innerWidth -= padding.left + padding.right
        }
        val lines: List<Line> = renderablesToLines(contents, innerWidth)
        var width = 0f
        var height = 0f
        for (line in lines) {
            width = max(width, line.width)
            height += line.height()
        }
        widthCache.put(maxWidth, PreCalcLines(lines, XyDim(width, height)))
    }

    private fun ensurePreCalcLines(maxWidth: Float): PreCalcLines {
        var pcl: PreCalcLines? = widthCache[maxWidth]
        if (pcl == null) {
            calcDimensionsForReal(maxWidth)
            pcl = widthCache[maxWidth]
        }
        return pcl!!
    }

    override fun renderator(): CellRenderator {
        return CellRenderator(this)
    }


    /** {@inheritDoc}  */
    fun calcDimensions(maxWidth: Float): XyDim? {
        // I think zero or negative width cells might be OK to ignore.  I'd like to try to make
        // Text.calcDimensionsForReal() handle this situation before throwing an error here.
        //        if (maxWidth < 0) {
        //            throw new IllegalArgumentException("maxWidth must be positive, not " + maxWidth);
        //        }
        val blockDim = ensurePreCalcLines(maxWidth).totalDim
        return if (cellStyle.padding == null) blockDim else cellStyle.padding.addTo(blockDim)
        //        System.out.println("Cell.calcDimensions(" + maxWidth + ") dim=" + dim +
        //                           " returns " + ret);
    }

    // TODO!
    inner class CellRenderator(c: Cell) : Renderator {

        override fun hasMore(): Boolean {
            return false
        }

        override fun getSomething(maxWidth: Float): ContTerm? {
            return null
        }

        override fun getIfFits(remainingWidth: Float): ContTermNone? {
            return null
        }
    }

    /*
    Renders item and all child-lines with given width and returns the x-y pair of the
    lower-right-hand corner of the last line (e.g. of text).

    {@inheritDoc}
    */
    fun render(lp: RenderTarget, outerTopLeft: XyOffset, outerDimensions: XyDim): XyOffset {
        //        System.out.println("Cell.render(" + this.toString());
        //        new Exception().printStackTrace();

        val maxWidth:Float = outerDimensions.width
        val pcls = ensurePreCalcLines(maxWidth)
        val padding = cellStyle.padding
        // XyDim outerDimensions = padding.addTo(pcrs.dim);

        // Draw background first (if necessary) so that everything else ends up on top of it.
        if (cellStyle.bgColor != null) {
            //            System.out.println("\tCell.render calling putRect...");
            lp.fillRect(outerTopLeft, outerDimensions, cellStyle.bgColor)
            //            System.out.println("\tCell.render back from putRect");
        }

        // Draw contents over background, but under border
        var innerTopLeft: XyOffset
        val innerDimensions: XyDim
        if (padding == null) {
            innerTopLeft = outerTopLeft
            innerDimensions = outerDimensions
        } else {
            //            System.out.println("\tCell.render outerTopLeft before padding=" + outerTopLeft);
            innerTopLeft = padding.applyTopLeft(outerTopLeft)
            //            System.out.println("\tCell.render innerTopLeft after padding=" + innerTopLeft);
            innerDimensions = padding.subtractFrom(outerDimensions)
        }
        val wrappedBlockDim = pcls.totalDim
        //        System.out.println("\tCell.render cellStyle.align()=" + cellStyle.align());
        //        System.out.println("\tCell.render outerDimensions=" + outerDimensions);
        //        System.out.println("\tCell.render padding=" + padding);
        //        System.out.println("\tCell.render innerDimensions=" + innerDimensions);
        //        System.out.println("\tCell.render wrappedBlockDim=" + wrappedBlockDim);
        val alignPad = cellStyle.align.calcPadding(innerDimensions, wrappedBlockDim)
        //        System.out.println("\tCell.render alignPad=" + alignPad);
        innerTopLeft = XyOffset(innerTopLeft.x + alignPad.left,
                                innerTopLeft.y - alignPad.top)

        var outerLowerRight = innerTopLeft
        var y:Float = innerTopLeft.y
        for (line in pcls.lines) {
            val rowXOffset = cellStyle.align
                    .leftOffset(wrappedBlockDim.width, line.width)
            outerLowerRight = line.render(lp,
                                          XyOffset(rowXOffset + innerTopLeft.x, y))
            y -= line.height()
            //            innerTopLeft = outerLowerRight.x(innerTopLeft.x());
        }

        // Draw border last to cover anything that touches it?
        val border = cellStyle.borderStyle
        if (border != null) {
            val origX:Float = outerTopLeft.x
            val origY:Float = outerTopLeft.y
            val rightX:Float = outerTopLeft.x + outerDimensions.width

            // This breaks cell rows in order to fix rendering content after images that fall
            // mid-page-break.  Math.min() below is so that when the contents overflow the bottom
            // of the cell, we adjust the cell border downward to match.  We aren't doing the same
            // for the background color, or for the rest of the row, so that's going to look bad.
            //
            // To fix these issues, I think we need to make that adjustment in the pre-calc instead
            // of here.  Which means that the pre-calc needs to be aware of page breaking because
            // the code that causes this adjustment is PdfLayoutMgr.appropriatePage().  So we
            // probably need a fake version of that that doesn't cache anything for display on the
            // page, then refactor backward from there until we enter this code with pre-corrected
            // outerLowerRight and can get rid of Math.min.
            //
            // When we do that, we also want to check PageGrouping.drawJpeg() and .drawPng()
            // to see if `return y + pby.adj;` still makes sense.
            val bottomY = Math.min(outerTopLeft.y - outerDimensions.height,
                                   outerLowerRight.y)

            // Like CSS it's listed Top, Right, Bottom, left
            if (border.top != null) {
                lp.drawLine(origX, origY, rightX, origY, border.top)
            }
            if (border.right != null) {
                lp.drawLine(rightX, origY, rightX, bottomY, border.right)
            }
            if (border.bottom != null) {
                lp.drawLine(origX, bottomY, rightX, bottomY, border.bottom)
            }
            if (border.left != null) {
                lp.drawLine(origX, origY, origX, bottomY, border.left)
            }
        }

        return outerLowerRight
    }

    // Replaced with TableRow.CellBuilder()
    //    /**
    //     Be careful when adding multiple cell builders at once because the cell size is based upon
    //     a pointer into the list of cell sizes.  That pointer gets incremented each time a cell is
    //     added, not each time nextCellSize() is called.  Is this a bug?  Or would fixing it create
    //     too many other bugs?
    //     @param trb
    //     @return
    //     */
    //    public static Builder builder(TableRowBuilder trb) {
    //        Builder b = new Builder(trb.cellStyle(), trb.nextCellSize()).textStyle(trb.textStyle());
    //        b.trb = trb;
    //        return b;
    //    }

    /**
     * A mutable Builder for somewhat less mutable cells.
     */
    class Builder(var cellStyle: CellStyle?,
                  override val width: Float,
                  val rows:MutableList<Renderable>,
                  var textStyle: TextStyle?) : CellBuilder {
        constructor(cs:CellStyle, w:Float) : this(cs, w, mutableListOf(), null)

        // Is this necessary?
        //        public Builder width(float w) { width = w; return this; }

        /** {@inheritDoc}  */
        override fun cellStyle(cs: CellStyle): Builder {
            cellStyle = cs
            return this
        }

        /** {@inheritDoc}  */
        override fun align(align: CellStyle.Align): Builder {
            cellStyle = cellStyle?.align(align) ?: CellStyle(align, null, null, null)
            return this
        }

        /** {@inheritDoc}  */
        override fun textStyle(x: TextStyle): Builder {
            textStyle = x
            return this
        }

        /** {@inheritDoc}  */
        override fun add(rs: Renderable): Builder {
            rows.add(rs)
            return this
        }

        /** {@inheritDoc}  */
        override fun addAll(js: Collection<Renderable>): Builder {
            rows.addAll(js)
            return this
        }

        /** {@inheritDoc}  */
        override fun add(ts: TextStyle, ls: Iterable<String>): Builder {
            for (s in ls) {
                rows.add(Text(ts, s))
            }
            return this
        }

        /** {@inheritDoc}  */
        override fun addStrs(vararg ss: String): Builder {
            if (textStyle == null) {
                throw IllegalStateException("Must set a default text style before adding raw strings")
            }
            for (s in ss) {
                rows.add(Text(textStyle!!, s))
            }
            return this
        }
        //        public Builder add(Cell c) { contents.add(c); return this; }

        fun build(): Cell {
            return Cell(cellStyle!!, width, rows)
        }

        // Replaced with TableRow.CellBuilder.buildCell()
        //        public TableRowBuilder buildCell() {
        //            Cell c = new Cell(cellStyle, width, contents);
        //            return trb.addCell(c);
        //        }

        /** {@inheritDoc}  */
        override fun toString(): String {
            val sB = StringBuilder("Cell.Builder(").append(cellStyle).append(" width=")
                    .append(width).append(" contents=[")

            var i = 0
            while (i < rows.size && i < 3) {
                if (i > 0) {
                    sB.append(" ")
                }
                sB.append(rows[i])
                i++
            }
            return sB.append("])").toString()
        }
    }

    /** {@inheritDoc}  */
    override fun toString(): String {
        val sB = StringBuilder("Cell(").append(cellStyle).append(" width=")
                .append(width).append(" contents=[")

        var i = 0
        while (i < contents.size && i < 3) {
            if (i > 0) {
                sB.append(" ")
            }
            sB.append(contents[i])
            i++
        }
        return sB.append("])").toString()
    }

    companion object {

        /**
         * Creates a new cell with the given style and width.
         *
         * @param cs    the cell style
         * @param width the width (height will be calculated based on how objects can be rendered within
         * this width).
         * @return a cell suitable for rendering.
         */
        fun of(cs: CellStyle, width: Float): Cell { //, final Object... r) {
            return Cell(cs, width, emptyList())
            //                        (r == null) ? Collections.emptyList()
            //                                    : Arrays.asList(r));
        }

        // Simple case of a single styled String
        fun of(cs: CellStyle, width: Float, ts: TextStyle, s: String): Cell {
            val ls = ArrayList<Renderable>(1)
            ls.add(Text(ts, s))
            return Cell(cs, width, ls)
        }

        // Simple case of a single styled String
        fun of(cs: CellStyle, width: Float, t: Text): Cell {
            val ls = ArrayList<Renderable>(1)
            ls.add(t)
            return Cell(cs, width, ls)
        }

        fun of(cs: CellStyle, width: Float, j: ScaledJpeg): Cell {
            val ls = ArrayList<Renderable>(1)
            ls.add(j)
            return Cell(cs, width, ls)
        }

        fun of(cs: CellStyle, width: Float, r: Renderable): Cell {
            val ls = ArrayList<Renderable>(1)
            ls.add(r)
            return Cell(cs, width, ls)
        }

        fun of(cs: CellStyle, width: Float, ls: List<Renderable>): Cell {
            return Cell(cs, width, ls)
        }

        // Simple case of a single styled String
        fun of(cs: CellStyle, width: Float, c: Cell): Cell {
            val ls = ArrayList<Renderable>(1)
            ls.add(c)
            return Cell(cs, width, ls)
        }

        fun builder(cellStyle: CellStyle, width: Float): Builder {
            return Builder(cellStyle, width)
        }
    }
}
